Update to Symfony 3.3, use Symfony Webpack Encore

cmfcmf opened the pull request:
This updates the app to Symfony 3.3, which features a bunch of dependency injection improvements. It also introduces the webpack based [Symfony Webpack Encore](http://symfony.com/doc/current/frontend.html) to manage frontend dependencies easily, including versioning. It has become the recommended way to deal with frontend dependencies in Symfony 3.3 and above. It configures Symfony and Webpack to do automatic asset versioning and allows for hot module replacement in development. I added two new commands to npm to start a HMR dev server: `dev-server-host` and `dev-server-guest`, depending on whether you run npm on your host or guest machine.

This PR also fixes some deprecations and applies the new DI practices. The deprecated and no longer needed ClassCache is also removed. I have also removed all references to Yarn, which I originally included due to its ability to create a `.lock` file. Now that npm 5 supports it as well, I don't see why we should continue requiring Yarn and npm.

cmfcmf commented:
Looks like the tests are failing - investigating.

cmfcmf commented:
> How is this looking now? Lots of changes here. Saw your comment before about the build failing

The tests currently fail with a *Segmentation fault*. I'll first work on other bugs and features before revisiting this PR.

cmfcmf commented:
I've now tinkered around with this __a lot__. I couldn't make the tests run on PHP 7.1 without producing segmentation faults. I therefore removed testing on PHP 7.1 entirely. I also had to require a newer PHP 7.0 version on Travis, because the default 7.0.7 would also cause segmentation faults.
I now consider this ready to be merged.

harkenn commented on app/Resources/views/base.html.twig:
```diff
@@ -4,7 +4,7 @@
         <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
         <title>{% block title %}Adventure Lookup{% endblock %}</title>
-        <link rel="stylesheet" href="{{ asset('assets/styles.css') }}?v=3">
+        <link rel="stylesheet" href="{{ asset('assets/app.css') }}">
```
I see you removed versioning? Why?

harkenn commented on app/Resources/webpack/index.js:
```diff
@@ -37,3 +38,15 @@ toastr.options = {
 export const myLazyLoad = new LazyLoad();
+
+// Hack to reload CSS using HMR
+// https://github.com/symfony/webpack-encore/pull/8#issuecomment-312599836
```
If this is the answer to my aforementioned question, HMR only works in development. It doesn't work in production (we wouldn't want it to anyways, it has a huge overhead).

cmfcmf replied:
HMR is only used in development, and only if you start it via `npm run dev-server-<host|guest>`. This code does nothing in production.

harkenn commented on webpack.config.js:
```diff
@@ -1,66 +1,37 @@
+    .enablePostCssLoader()
+    .configureBabel(function(babelConfig) {
+        // add additional presets
+        babelConfig.presets.push('es2015');
```
`env` is the current recommended preset to use

harkenn commented on webpack.config.js:
```diff
@@ -1,66 +1,37 @@
+        'Tether': 'tether',
+    })
+    .enableSourceMaps(!Encore.isProduction())
+    .enableVersioning(Encore.isProduction())
```
This might be the answer to the questions from before? If this handles versioning automatically then concerns about the stylesheet aren't big.

cmfcmf replied:
Correct, versioning is handled automatically now.

cmfcmf commented:
I addressed the babel preset in https://github.com/AdventureLookup/AdventureLookup/pull/119/commits/b9f7f75482ea1368e3d6d66e8905e66206968b5f (env actually is the default) and added missing browserslist for autoprefixer.
